### PR TITLE
relayer: Catches executeInstruction reverts & Deals with bad reward addresses

### PR DESCRIPTION
**Pull request by derpy-duck**
- Only decodes the revert message from executeInstruction if it matches one of three known error types
- Revert when ‘pay()’ fails

### CODE REVIEW

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 1]
```diff
@@ -488,17 +502,17 @@ abstract contract CoreRelayerDelivery is CoreRelayerBase, IWormholeRelayerDelive
...
   }
 
-  function decodeCancelled(
+  function decodeExecuteInstructionError(
     bytes memory revertData
-  ) private pure returns (uint32 gasUsed, uint256 available, uint256 required) {
+  ) private pure returns (uint32 gasUsed, bool knownError) {
     uint offset = 0;
     bytes4 selector;
     (selector, offset) = revertData.asBytes4Unchecked(offset);
```
Comment (scnale) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:510]: Do we know that the buffer has at least 4 bytes at this point? We might want to check for the entire exact size or a lower bound for it.

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 2]
```diff
@@ -488,17 +502,17 @@ abstract contract CoreRelayerDelivery is CoreRelayerBase, IWormholeRelayerDelive
     }
   }
 
-  function decodeCancelled(
+  function decodeExecuteInstructionError(
```
Comment (scnale) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:505]: Perhaps we could name this `tryDecodeExecuteInstructionError`.

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 1 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:514]: Ah good point. Fixed!

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 2 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol]: done

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 1 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:514]: Ah yes oops, corrected!

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 3]
```diff
@@ -356,6 +358,10 @@ abstract contract CoreRelayerDelivery is CoreRelayerBase, IWormholeRelayerDelive
     Wei transactionFeeRefundAmount,
     ForwardInstruction[] storage forwardInstructions
   ) private {
+     //despite being external, we only allow ourselves to call this function (via CALL opcode)
+    //  used as a means to retroactively revert the call if the calls to the relay provider here revert
+    if (msg.sender != address(this))
+      revert RequesterNotCoreRelayer();
```
Comment (scnale) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:364]: Should this be placed here? This is a private function.

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol [hunk 4]
```diff
@@ -488,17 +502,21 @@ abstract contract CoreRelayerDelivery is CoreRelayerBase, IWormholeRelayerDelive
...
     (selector, offset) = revertData.asBytes4Unchecked(offset);
-    offset += 28;
-    (gasUsed, offset) = revertData.asUint32Unchecked(offset);
-    (available, offset) = revertData.asUint256Unchecked(offset);
-    (required, offset) = revertData.asUint256Unchecked(offset);
-    assert(offset == revertData.length && selector == Cancelled.selector);
+    if((selector == Cancelled.selector) || (selector == RelayProviderReverted.selector) || (selector == RelayProviderPaymentFailed.selector)) {
+      knownError = true;
+      offset += 28;
+      (gasUsed, offset) = revertData.asUint32Unchecked(offset);
```
Comment (scnale) [inline ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:518]: I think we should have a bounds check for this too. I think I'd just unify the bounds check with the above. We would check that the `revertData` has at least 36 bytes; enough to hold the selector and the ABI encoded `uint32` argument.

Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol:516]: agreed!

File: ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol (hunk 3 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol]: ah good catch, fixed

**Review by scnale — APPROVED** [review]

**Pull request merged**